src/heap.cpp

```diff
@@ -1,23 +1,218 @@
+template<class ValueType> void heapMode(int heapType);
+template<class ValueType> void sortMode(int heapType);
+
+template<class ValueType> void crateHeap(Heap<ValueType>** heap, int heapType);
+template<class ValueType> void insertToHeap(Heap<ValueType>* heap);
+template<class ValueType> void getFromHeap(Heap<ValueType>* heap);
+template<class ValueType> void getArray(vector <ValueType>& newValues);
+
+
 int main()
 {
-	Heap<int>* testHeap = new MaxHeap<int>;
-	testHeap->insertValue(1);
-	testHeap->insertValue(2);
-	testHeap->insertValue(3);
-	testHeap->insertValue(4);
-	testHeap->insertValue(5);
-	testHeap->insertValue(6);
-	testHeap->insertValue(7);
-	testHeap->insertValue(8);
-	testHeap->insertValue(9);
-	testHeap->insertValue(10);
-
-	testHeap->show();
-	delete testHeap;
+	int mode = chooseMode();
+	int heapType, valueType;
+	chooseHeapProperties(heapType, valueType);
+
+	if (mode == 1) {
+		switch (valueType) {
+		case 3:
+			heapMode<string>(heapType);
+			break;
+		case 2:
+			heapMode<float>(heapType);
+			break;
+		default:
+			heapMode<int>(heapType);
+			break;
+		}
+	}
+	else {
+		switch (valueType) {
+		case 3: 
+			sortMode<string>(heapType);
+			break;
+		case 2:
+			sortMode<float>(heapType);
+			break;
+		default: 
+			sortMode<int>(heapType);
+			break;
+		}
+	}
+}
+
+
+template<class ValueType>
+void getValidValue(ValueType& value) {
+	cin.clear();
+	cin.sync();
+	cin >> value;
+
+	while (cin.fail()) {
+		getchar();
+		cin.clear();
+		cin.sync();
+		cout << "Niepoprawny wybor. Sprobuj jeszcze raz: ";
+		cin >> value;
+	}
+}
+
+template<class ValueType>
+void getValueFromRange(ValueType& value, int min, int max) {
+	getValidValue(value);
+	while (value < min || value > max) {
+		cout << "Niepoprawny wybor. Sprobuj jeszcze raz: ";
+		getValidValue(value);
+	}
+}
+
+int chooseMode() {
+	int choice;
+	cout << "--KOPIEC--" << endl;
+	cout << "1. Dodawanie i usuwanie wartosci z kopca" << endl;
+	cout << "2. Posortuj z uzyciem kopca" << endl;
+	cout << "Wybierz opcje (1-2): ";
+	getValueFromRange(choice, 1, 2);
+	system("CLS");
+	return choice;
+}
+
+void chooseHeapProperties(int& heapType, int& valueType) {
+	cout << "Wybierz parametry dla nowego kopca" << endl;
+	cout << "I Typ kopca: " << endl;
+	cout << "1. Kopiec typu max" << endl;
+	cout << "2. Kopiec typu min" << endl;
+	cout << "Podaj typ (1-2): ";
+	getValueFromRange(heapType, 1, 2);
+	cout << endl;
+
+	cout << "II Wartosci w kopcu: " << endl;
+	cout << "1. Liczby calkowite" << endl;
+	cout << "2. Liczby zmiennoprzecinkowe" << endl;
+	cout << "3. Slowa" << endl;
+	cout << "Podaj typ (1-3): ";
+	getValueFromRange(valueType, 1, 3);
+	system("CLS");
+}
+
+int showMenu() {
+	int choice;
+	cout << "--MENU--" << endl;
+	cout << "1. Dodaj" << endl;
+	cout << "2. Usun" << endl;
+	cout << "3. Zwroc" << endl;
+	cout << "4. Wyjdz" << endl;
+	cout << "Wybierz opcje (1-4): ";
+	getValueFromRange(choice, 1, 4);
+	system("CLS");
+	return choice;
 }
+
+template<class ValueType>
```

Comment:
W C++ trzeba przekazywać tego Heap<ValueType>** heap jako paramter metody, czy można by było po prosty zwrócić ten wskażnik w return? Jak tak to w samej metodzie też pewnie można by było użyć '? :'